lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+    if (!site) {
+      return "";
+    }
+
+    let queryParams = new Map();
+    let safeQueryParam = "";
+    let url = new URL(site);
+
+    if (url.search.length === 0) {
+      return url.toString();
+    }
+    // extract and parse the query parameters, if any
+    url.search.slice(1, url.search.length);
+    let paramArray = url.search.slice(1, url.search.length).split("&");
+    paramArray.forEach(item => {
+      let tmp = item.split("=");
+      queryParams.set(tmp[0], tmp[1]);
+    });
+    // remove the unwanted parameters and update the query string
+    REMOVE_KEYS.forEach(key => queryParams.delete(key));
+    queryParams.forEach((val, key) => {
+      if (val.length !== 0 && key.length !== 0) {
+        safeQueryParam = safeQueryParam + key + "=" + val + "&";
+      }
+    });
+    url.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    url.hash = "";
+
+    // remove extra slashes in pathname and construct back a safe pathname
+    if (url.pathname) {
+      let pathItems = url.pathname.split("/");
+      let safePathItems = pathItems.filter(item => item.length !== 0);
+      let safePath = "/";
+      url.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url doesn't have a protocol, use https
+    if (!url.protocol) {
+      url.protocol = "https:";
+    }
+
+    return url.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key;
+  },
+
+  /**
+    * Canonicalize urls by deduping them, then sanitizing them
+    */
+  _dedupeURLs(sites) {
+    let dedupedSites = new Map();
+    sites.forEach(site => {
+      let urlkey = this._createKey(site.url);
+      if (!dedupedSites.has(urlkey)) {
+        dedupedSites.set(urlkey, site);
+      } else {
+        // if two sites come in with 'http' and 'https', pick 'https'
+        if (!dedupedSites.get(urlkey).url.contains("https")) {
+          dedupedSites.delete(urlkey);
+          dedupedSites.set(urlkey, site);
+        }
+      }
+    });
+    let freshSites = [];
+    dedupedSites.forEach((site) => {
+      // sanitize and normalize each site
+      site.url = this._sanitizeURL(site.url);
+      freshSites.push(site);
+    });
+    return freshSites;
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    let sites = this._dedupeURLs(links);
+    sites.forEach(link => {
+      if (link && ss.storage.embedlyData[link.url] && link.url === ss.storage.embedlyData[link.url].url) {
+        cachedLinks.push(ss.storage.embedlyData[link.url]);
+      }
+    });
+    return cachedLinks;
+  },
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let newLinks = links.filter(link => link && !ss.storage.embedlyData[link.url]);
```

Comment:
@oyiptong @sarracini I think we can simplify this using the URL layers described above

1) Receive a list of URLs from the places db

2) Sanitize them using the above method

3) Passing that sanitized list through a set will remove duplicates

4) Generate the cache keys of the sanitized URLs to look up in the simple storage local cache

5) Find the set of URLs which were not locally cached and pass those to the embedly proxy (these are sanitized, de duped, and not locally cached)